datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomDataMapWriter.java

```diff
@@ -21,142 +21,124 @@
-
   }
 
   @Override
   public void onBlockletStart(int blockletId) {
-    this.currentBlockletId = blockletId;
+  }
+
+  protected void resetBloomFilters() {
     indexBloomFilters.clear();
-    for (int i = 0; i < indexedColumns.size(); i++) {
+    List<CarbonColumn> indexColumns = getIndexColumns();
+    for (int i = 0; i < indexColumns.size(); i++) {
       indexBloomFilters.add(BloomFilter.create(Funnels.byteArrayFunnel(),
           bloomFilterSize, 0.00001d));
     }
   }
 
   @Override
   public void onBlockletEnd(int blockletId) {
-    try {
-      writeBloomDataMapFile();
-    } catch (Exception e) {
-      for (ObjectOutputStream objectOutputStream : currentObjectOutStreams) {
-        CarbonUtil.closeStreams(objectOutputStream);
-      }
-      for (DataOutputStream dataOutputStream : currentDataOutStreams) {
-        CarbonUtil.closeStreams(dataOutputStream);
-      }
-      throw new RuntimeException(e);
-    }
+    writeBloomDataMapFile();
+    currentBlockletId++;
   }
 
-  // notice that the input pages only contains the indexed columns
   @Override
-  public void onPageAdded(int blockletId, int pageId, ColumnPage[] pages)
-      throws IOException {
-    col2Ordianl.clear();
-    col2DataType.clear();
-    for (int colId = 0; colId < pages.length; colId++) {
-      String columnName = pages[colId].getColumnSpec().getFieldName().toLowerCase();
-      col2Ordianl.put(columnName, colId);
-      DataType columnType = pages[colId].getColumnSpec().getSchemaDataType();
-      col2DataType.put(columnName, columnType);
-    }
+  public void addRow(int blockletId, int pageId, int rowId, CarbonRow row) {
+    addRow(row.getData());
+  }
 
-    // for each row
-    for (int rowId = 0; rowId < pages[0].getPageSize(); rowId++) {
-      // for each indexed column
-      for (int indexColId = 0; indexColId < indexedColumns.size(); indexColId++) {
-        String indexedCol = indexedColumns.get(indexColId);
-        byte[] indexValue;
-        if (DataTypes.STRING == col2DataType.get(indexedCol)
-            || DataTypes.BYTE_ARRAY == col2DataType.get(indexedCol)) {
-          byte[] originValue = (byte[]) pages[col2Ordianl.get(indexedCol)].getData(rowId);
-          indexValue = new byte[originValue.length - 2];
-          System.arraycopy(originValue, 2, indexValue, 0, originValue.length - 2);
-        } else {
-          Object originValue = pages[col2Ordianl.get(indexedCol)].getData(rowId);
-          indexValue = CarbonUtil.getValueAsBytes(col2DataType.get(indexedCol), originValue);
-        }
-
-        indexBloomFilters.get(indexColId).put(indexValue);
+  protected void addRow(Object[] rowData) {
+    // for each indexed column, add the data to bloom filter
+    List<CarbonColumn> indexColumns = getIndexColumns();
+    for (int i = 0; i < indexColumns.size(); i++) {
+      DataType dataType = indexColumns.get(i).getDataType();
+      byte[] indexValue;
+      if (DataTypes.STRING == dataType) {
```

Comment:
The procedure for string and byte_array are the same, there is no need to distinguish them.

Reply:
getStringData need to be override by subclass (BloomDataMapRefresher), BloomDataMapRefresher will get the row data from IndexDataMapRefreshRDD.internalCompute, it needs a different implementation